lfs/gitscanner_index.go

```diff
@@ -106,67 +105,46 @@ func scanIndex(cb GitScannerFoundPointer, ref string) error {
 // for in the indexf. It returns a channel from which sha1 strings can be read.
 // The namMap will be filled indexFile pointers mapping sha1s to indexFiles.
 func revListIndex(atRef string, cache bool, indexMap *indexFileMap) (*StringChannelWrapper, error) {
-	cmdArgs := []string{"diff-index", "-M"}
-	if cache {
-		cmdArgs = append(cmdArgs, "--cached")
-	}
-	cmdArgs = append(cmdArgs, atRef)
-
-	cmd, err := startCommand("git", cmdArgs...)
+	scanner, cmd, err := NewDiffIndexScanner(atRef, cache)
 	if err != nil {
 		return nil, err
 	}
 
-	cmd.Stdin.Close()
-
 	revs := make(chan string, chanBufSize)
-	errchan := make(chan error, 1)
+	errs := make(chan error, 1)
 
 	go func() {
-		scanner := bufio.NewScanner(cmd.Stdout)
 		for scanner.Scan() {
-			// Format is:
-			// :100644 100644 c5b3d83a7542255ec7856487baa5e83d65b1624c 9e82ac1b514be060945392291b5b3108c22f6fe3 M foo.gif
-			// :<old mode> <new mode> <old sha1> <new sha1> <status>\t<file name>[\t<file name>]
-			line := scanner.Text()
-			parts := strings.Split(line, "\t")
-			if len(parts) < 2 {
-				continue
+			var name string = scanner.Entry().DstName
+			if len(name) == 0 {
+				name = scanner.Entry().SrcName
 			}
 
-			description := strings.Split(parts[0], " ")
-			files := parts[1:len(parts)]
+			var sha string = scanner.Entry().DstSha
+			if scanner.Entry().Status == StatusModification {
+				sha = scanner.Entry().SrcSha
+			}
 
-			if len(description) >= 5 {
-				status := description[4][0:1]
-				sha1 := description[3]
-				if status == "M" {
-					sha1 = description[2] // This one is modified but not added
-				}
+			indexMap.Add(sha, &indexFile{
```

Comment:
This `*indexFile` is kept over the new `DiffIndexEntry` type, since the `Name` field (and `sha` value) aren't entirely the output from the `diff-index` command.

- `Name` refers to either the new name of the file (if renamed), or the `SrcName` field.
- `Sha` refers to either the `DstSha`, or the `SrcSha`, if the status is 'M' for modified.